src/NServiceBus.Core/Unicast/RunningEndpointInstance.cs

```diff
@@ -63,15 +65,23 @@ public async Task Stop(CancellationToken cancellationToken = default)
                 finally
                 {
                     settings.Clear();
-                    (builder as IDisposable)?.Dispose();
+                    if (builder is IAsyncDisposable asyncDisposableBuilder)
+                    {
+                        await asyncDisposableBuilder.DisposeAsync().ConfigureAwait(false);
+                    }
```

Comment:
Is it safer if we still fall back to check for `IDisposable`?  `ServiceProvider` may be `IAsyncDisposable` but other implementations of `IServiceProvider` may not be, and I don't know if it's possible that another implementation is used.

Reply:
That could be true for external builders. Here this code is only ever executed when `shouldDisposeBuilder` is `true` which is only the case when NServiceBus manages the container (internal managed mode). In this case it is always the `ServiceProvider`

https://github.com/Particular/NServiceBus/blob/master/src/NServiceBus.Core/Endpoint.cs#L23-L25

I don't like how the code is structured. Currently, it propagates the `shouldDisposeBuilder` forward to a certain point and then replaces the false case by passing `null` and the `true` case by passing the `IServiceProvider` instance, which is always the internally managed. But refactoring this would bit a bit of an undertaking potentially. 

So the minimum thing I did was enabling nullable on this type to indicate that some constructor arguments might be null. 

In case you have a good suggestion how to make this simpler I'm all ears.

Reply:
I've pushed some changes to make things more revealing (I hope)